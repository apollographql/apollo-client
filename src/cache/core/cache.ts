import { DocumentNode } from 'graphql';
import { wrap } from 'optimism';

import { getFragmentQueryDocument } from '../../utilities/graphql/fragments';
import { StoreObject } from '../../utilities/graphql/storeUtils';
import { DataProxy } from './types/DataProxy';
import { Cache } from './types/Cache';
import { Modifier, Modifiers } from './types/common';

export type Transaction<T> = (c: ApolloCache<T>) => void;

export abstract class ApolloCache<TSerialized> implements DataProxy {
  // required to implement
  // core API
  public abstract read<T, TVariables = any>(
    query: Cache.ReadOptions<TVariables>,
  ): T | null;
  public abstract write<TResult = any, TVariables = any>(
    write: Cache.WriteOptions<TResult, TVariables>,
  ): void;
  public abstract diff<T>(query: Cache.DiffOptions): Cache.DiffResult<T>;
  public abstract watch(watch: Cache.WatchOptions): () => void;
  public abstract reset(): Promise<void>;

  // If called with only one argument, removes the entire entity
  // identified by dataId. If called with a fieldName as well, removes all
  // fields of the identified entity whose store names match fieldName.
  public abstract evict(dataId: string, fieldName?: string): boolean;

  // intializer / offline / ssr API
  /**
   * Replaces existing state in the cache (if any) with the values expressed by
   * `serializedState`.
   *
   * Called when hydrating a cache (server side rendering, or offline storage),
   * and also (potentially) during hot reloads.
   */
  public abstract restore(
    serializedState: TSerialized,
  ): ApolloCache<TSerialized>;

  /**
   * Exposes the cache's complete state, in a serializable format for later restoration.
   */
  public abstract extract(optimistic?: boolean): TSerialized;

  // Optimistic API

  public abstract removeOptimistic(id: string): void;

  // Transactional API

  public abstract performTransaction(
    transaction: Transaction<TSerialized>,
  ): void;

  public abstract recordOptimisticTransaction(
    transaction: Transaction<TSerialized>,
    id: string,
  ): void;

  // Optional API

  public transformDocument(document: DocumentNode): DocumentNode {
    return document;
  }

  public identify(object: StoreObject): string | undefined {
    return;
  }

  public modify(
    modifiers: Modifier<any> | Modifiers,
    dataId?: string,
    optimistic?: boolean,
  ): boolean {
    return false;
  }
  
  public gc(): string[] {
    return [];
  }

  // Experimental API

  public transformForLink(document: DocumentNode): DocumentNode {
    return document;
  }

  // DataProxy API
  /**
   *
   * @param options
   * @param optimistic
   */
  public readQuery<QueryType, TVariables = any>(
    options: DataProxy.Query<TVariables>,
    optimistic: boolean = false,
  ): QueryType | null {
    return this.read({
      rootId: options.id || 'ROOT_QUERY',
      query: options.query,
      variables: options.variables,
      optimistic,
    });
  }

  // Make sure we compute the same (===) fragment query document every
  // time we receive the same fragment in readFragment.
  private getFragmentDoc = wrap(getFragmentQueryDocument);

  public readFragment<FragmentType, TVariables = any>(
    options: DataProxy.Fragment<TVariables>,
    optimistic: boolean = false,
  ): FragmentType | null {
    return this.read({
      query: this.getFragmentDoc(options.fragment, options.fragmentName),
      variables: options.variables,
      rootId: options.id,
      optimistic,
    });
  }

  public writeQuery<TData = any, TVariables = any>(
    options: Cache.WriteQueryOptions<TData, TVariables>,
  ): void {
    this.write({
      dataId: options.id || 'ROOT_QUERY',
      result: options.data,
      query: options.query,
      variables: options.variables,
    });
  }

  public writeFragment<TData = any, TVariables = any>(
    options: Cache.WriteFragmentOptions<TData, TVariables>,
  ): void {
    this.write({
      dataId: options.id,
      result: options.data,
      variables: options.variables,
      query: this.getFragmentDoc(options.fragment, options.fragmentName),
    });
  }
}
