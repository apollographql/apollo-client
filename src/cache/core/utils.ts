import {
  DocumentNode,
  SelectionSetNode,
} from 'graphql';

export function queryFromPojo(obj: any): DocumentNode {
  return {
    kind: 'Document',
    definitions: [{
      kind: 'OperationDefinition',
      operation: 'query',
      name: {
        kind: 'Name',
        value: 'GeneratedClientQuery',
      },
      selectionSet: selectionSetFromObj(obj),
    }],
  };
}

export function fragmentFromPojo(obj: any, typename?: string): DocumentNode {
  return {
    kind: 'Document',
    definitions: [{
      kind: 'FragmentDefinition',
      typeCondition: {
        kind: 'NamedType',
        name: {
          kind: 'Name',
          value: typename || '__FakeType',
        },
      },
      name: {
        kind: 'Name',
        value: 'GeneratedClientQuery',
      },
      selectionSet: selectionSetFromObj(obj),
    }],
  };
}

function selectionSetFromObj(obj: any): SelectionSetNode {
  if (!obj || Object(obj) !== obj) {
    // No selection set here
    return null;
  }

  if (Array.isArray(obj)) {
    // GraphQL queries don't include arrays
    return selectionSetFromObj(obj[0]);
  }

  // Now we know it's an object
  return {
    kind: 'SelectionSet',
    selections: Object.keys(obj).map(key => ({
      kind: 'Field',
      name: {
        kind: 'Name',
        value: key,
      },
      selectionSet: selectionSetFromObj(obj[key]) || void 0,
    })),
  };
}
