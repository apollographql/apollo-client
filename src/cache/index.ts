export { Transaction, ApolloCache } from './core/cache';
export { Cache } from './core/types/Cache';
export { DataProxy } from './core/types/DataProxy';
export { MissingFieldError } from './core/types/common';

export {
  Reference,
  isReference,
  makeReference,
} from '../utilities/graphql/storeUtils';

export {
  InMemoryCache,
  InMemoryCacheConfig,
  ReactiveVar,
} from './inmemory/inMemoryCache';

export {
  defaultDataIdFromObject,
  TypePolicies,
  TypePolicy,
  FieldPolicy,
  FieldReadFunction,
  FieldMergeFunction,
  FieldFunctionOptions,
  PossibleTypesMap,
} from './inmemory/policies';

export * from './inmemory/types';
