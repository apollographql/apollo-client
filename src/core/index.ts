/* Core */

export {
  ApolloClient,
  ApolloClientOptions,
  DefaultOptions,
  mergeOptions,
} from './ApolloClient';
export {
  ObservableQuery,
  FetchMoreOptions,
  UpdateQueryOptions,
} from './ObservableQuery';
export {
  QueryOptions,
  WatchQueryOptions,
  MutationOptions,
  SubscriptionOptions,
  FetchPolicy,
  WatchQueryFetchPolicy,
  ErrorPolicy,
  FetchMoreQueryOptions,
  SubscribeToMoreOptions,
  MutationUpdaterFn,
} from './watchQueryOptions';
export { NetworkStatus } from './networkStatus';
export * from './types';
export {
  Resolver,
  FragmentMatcher,
} from './LocalState';
export { isApolloError, ApolloError } from '../errors';

/* Cache */

export {
  // All the exports (types and values) from ../cache, minus cacheSlot,
  // which we want to keep semi-private.
  Cache,
  ApolloCache,
  Transaction,
  DataProxy,
  InMemoryCache,
  InMemoryCacheConfig,
  MissingFieldError,
  defaultDataIdFromObject,
  ReactiveVar,
  makeVar,
  TypePolicies,
  TypePolicy,
  FieldPolicy,
  FieldReadFunction,
  FieldMergeFunction,
  FieldFunctionOptions,
  PossibleTypesMap,
} from '../cache';

export * from '../cache/inmemory/types';

/* Link */

export * from '../link/core';
export * from '../link/http';
export {
  fromError,
  toPromise,
  fromPromise,
  ServerError,
  throwServerError,
} from '../link/utils';

/* Utilities */

export {
  Observable,
  Observer,
  ObservableSubscription,
  Reference,
  isReference,
  makeReference,
  StoreObject,
} from '../utilities';

/* Supporting */

// The verbosity of invariant.{log,warn,error} can be controlled globally
// (for anyone using the same ts-invariant package) by passing "log",
// "warn", "error", or "silent" to setVerbosity. By default, Apollo Client
// displays warnings and errors, but hides invariant.log statements. Note
// that all invariant.* logging is hidden in production.
import { setVerbosity } from "ts-invariant";
export { setVerbosity as setLogVerbosity }
setVerbosity("warn");

// Note that importing `gql` by itself, then destructuring
// additional properties separately before exporting, is intentional.
// Due to the way the `graphql-tag` library is setup, certain bundlers
// can't find the properties added to the exported `gql` function without
// additional guidance (e.g. Rollup - see
// https://rollupjs.org/guide/en/#error-name-is-not-exported-by-module).
// Instead of having people that are using bundlers with `@apollo/client` add
// extra bundler config to help `graphql-tag` exports be found (which would be
// awkward since they aren't importing `graphql-tag` themselves), this
// workaround of pulling the extra properties off the `gql` function,
// then re-exporting them separately, helps keeps bundlers happy without any
// additional config changes.
import gql from 'graphql-tag';
export const {
  resetCaches,
  disableFragmentWarnings,
  enableExperimentalFragmentVariables,
  disableExperimentalFragmentVariables
} = gql;
export { gql };
