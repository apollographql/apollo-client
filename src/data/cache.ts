import { DocumentNode } from 'graphql';
import {
  DataProxy,
  DataProxyReadQueryOptions,
  DataProxyReadFragmentOptions,
  DataProxyWriteQueryOptions,
  DataProxyWriteFragmentOptions,
} from './proxy';
import { addTypenameToDocument } from '../queries/queryTransform';
import {
  getFragmentQueryDocument,
  getOperationName,
} from '../queries/getFromAST';

export type CacheWrite = {
  dataId: string;
  result: any;
  document: DocumentNode;
  variables?: Object;
};

export abstract class Cache implements DataProxy {
  private addTypename: boolean;
  constructor(addTypename: boolean) {
    this.addTypename = addTypename;
  }

  public abstract reset(): Promise<void>;

  public abstract diffQuery(query: {
    query: DocumentNode;
    variables: any;
    returnPartialData?: boolean;
    previousResult?: any;
    optimistic: boolean;
  }): any;

  public abstract read(query: {
    query: DocumentNode;
    variables: any;
    rootId?: string;
    previousResult?: any;
    optimistic: boolean;
  }): any;

  public readQuery<QueryType>(
    options: DataProxyReadQueryOptions,
    optimistic: boolean = false,
  ): QueryType {
    let query = options.query;
    if (this.addTypename) {
      query = addTypenameToDocument(query);
    }

    return this.read({
      query,
      variables: options.variables,
      optimistic,
    });
  }

  public readFragment<FragmentType>(
    options: DataProxyReadFragmentOptions,
    optimistic: boolean = false,
  ): FragmentType | null {
    let document = getFragmentQueryDocument(
      options.fragment,
      options.fragmentName,
    );
    if (this.addTypename) {
      document = addTypenameToDocument(document);
    }

    return this.read({
      query: document,
      variables: options.variables,
      rootId: options.id,
      optimistic,
    });
  }

  public abstract writeResult(write: CacheWrite): void;

  public writeQuery(options: DataProxyWriteQueryOptions): void {
    let query = options.query;
    if (this.addTypename) {
      query = addTypenameToDocument(query);
    }

    this.writeResult({
      dataId: 'ROOT_QUERY',
      result: options.data,
      document: query,
      variables: options.variables,
    });
  }

  public writeFragment(options: DataProxyWriteFragmentOptions): void {
    let document = getFragmentQueryDocument(
      options.fragment,
      options.fragmentName,
    );
    if (this.addTypename) {
      document = addTypenameToDocument(document);
    }

    this.writeResult({
      dataId: options.id,
      result: options.data,
      document,
      variables: options.variables,
    });
  }

  public abstract removeOptimistic(id: string): void;

  public abstract performTransaction(transaction: (c: Cache) => void): void;
  public abstract recordOptimisticTransaction(
    transaction: (c: Cache) => void,
    id: string,
  ): void;
}
