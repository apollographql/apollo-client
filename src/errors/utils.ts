export function isBranded(error: unknown, name: string) {
  return (
    typeof error === "object" &&
    error !== null &&
    (error as any)[Symbol.for("apollo.error")] === name
  );
}

export function brand(error: Error) {
  Object.defineProperty(error, Symbol.for("apollo.error"), {
    value: error.name,
    enumerable: false,
    writable: false,
    configurable: false,
  });
}
