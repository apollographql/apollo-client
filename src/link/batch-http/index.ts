export { BatchHttpLink } from "./batchHttpLink.js";
