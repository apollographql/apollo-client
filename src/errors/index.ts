import '../utilities/globals';

import { GraphQLError, GraphQLErrorExtensions } from 'graphql';

import { isNonNullObject } from '../utilities';
import { ServerParseError } from '../link/http';
import { ServerError } from '../link/utils';
import { FetchResult } from "../link/core";

// This Symbol allows us to pass transport-specific errors from the link chain
// into QueryManager/client internals without risking a naming collision within
// extensions (which implementers can use as they see fit).
export const PROTOCOL_ERRORS_SYMBOL: unique symbol = Symbol();

type FetchResultWithSymbolExtensions<T> = FetchResult<T> & {
  extensions: Record<string | symbol, any>
};

export interface ApolloErrorOptions {
  graphQLErrors?: ReadonlyArray<GraphQLError>;
  protocolErrors?: ReadonlyArray<{
    message: string;
    extensions?: GraphQLErrorExtensions[];
  }>;
  clientErrors?: ReadonlyArray<Error>;
  networkError?: Error | ServerParseError | ServerError | null;
  errorMessage?: string;
  extraInfo?: any;
}

export function graphQLResultHasProtocolErrors<T>(
  result: FetchResult<T>
): result is FetchResultWithSymbolExtensions<T> {
  if (result.extensions) {
    return Array.isArray(
      (result as FetchResultWithSymbolExtensions<T>).extensions[
        PROTOCOL_ERRORS_SYMBOL
      ]
    );
  }
  return false;
}


export function isApolloError(err: Error): err is ApolloError {
  return err.hasOwnProperty('graphQLErrors');
}

// Sets the error message on this error according to the
// the GraphQL and network errors that are present.
// If the error message has already been set through the
// constructor or otherwise, this function is a nop.
const generateErrorMessage = (err: ApolloError) => {
  const errors = [
    ...err.graphQLErrors,
    ...err.clientErrors,
    ...err.protocolErrors
  ];
  if (err.networkError) errors.push(err.networkError);
  return errors
    // The rest of the code sometimes unsafely types non-Error objects as GraphQLErrors
    .map(err => isNonNullObject(err) && err.message || 'Error message not found.')
    .join('\n');
};

export type GraphQLErrors = ReadonlyArray<GraphQLError>;

export type NetworkError = Error | ServerParseError | ServerError | null;

export class ApolloError extends Error {
  public name: string;
  public message: string;
  public graphQLErrors: GraphQLErrors;
  public protocolErrors: ReadonlyArray<{
    message: string;
    extensions?: GraphQLErrorExtensions[];
  }>;
  public clientErrors: ReadonlyArray<Error>;
  public networkError: Error | ServerParseError | ServerError | null;

  // An object that can be used to provide some additional information
  // about an error, e.g. specifying the type of error this is. Used
  // internally within Apollo Client.
  public extraInfo: any;

  // Constructs an instance of ApolloError given a GraphQLError
  // or a network error. Note that one of these has to be a valid
  // value or the constructed error will be meaningless.
  constructor({
    graphQLErrors,
    protocolErrors,
    clientErrors,
    networkError,
    errorMessage,
    extraInfo,
  }: ApolloErrorOptions) {
    super(errorMessage);
    this.name = 'ApolloError';
    this.graphQLErrors = graphQLErrors || [];
    this.protocolErrors = protocolErrors || [];
    this.clientErrors = clientErrors || [];
    this.networkError = networkError || null;
    this.message = errorMessage || generateErrorMessage(this);
    this.extraInfo = extraInfo;

    // We're not using `Object.setPrototypeOf` here as it isn't fully
    // supported on Android (see issue #3236).
    (this as any).__proto__ = ApolloError.prototype;
  }
}
