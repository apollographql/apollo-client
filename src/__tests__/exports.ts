import * as cache from "../cache";
import * as client from "..";
import * as core from "../core";
import * as errors from "../errors";
import * as linkBatch from "../link/batch";
import * as linkBatchHTTP from "../link/batch-http";
import * as linkContext from "../link/context";
import * as linkCore from "../link/core";
import * as linkError from "../link/error";
import * as linkHTTP from "../link/http";
import * as linkPersistedQueries from "../link/persisted-queries";
import * as linkRetry from "../link/retry";
import * as linkSchema from "../link/schema";
import * as linkUtils from "../link/utils";
import * as linkWS from "../link/ws";
import * as react from "../react";
import * as reactComponents from "../react/components";
import * as reactContext from "../react/context";
import * as reactData from "../react/data";
import * as reactHOC from "../react/hoc";
import * as reactHooks from "../react/hooks";
import * as reactParser from "../react/parser";
import * as reactSSR from "../react/ssr";
import * as testing from "../testing";
import * as utilities from "../utilities";

const entryPoints = require("../../config/entryPoints.js");

type Namespace = object;

describe('exports of public entry points', () => {
  const testedIds = new Set<string>();

  function check(id: string, ns: Namespace) {
    it(id, () => {
      testedIds.add(id);
      expect(Object.keys(ns).sort()).toMatchSnapshot();
    });
  }

  check("@apollo/client", client);
  check("@apollo/client/cache", cache);
  check("@apollo/client/core", core);
  check("@apollo/client/errors", errors);
  check("@apollo/client/link/batch", linkBatch);
  check("@apollo/client/link/batch-http", linkBatchHTTP);
  check("@apollo/client/link/context", linkContext);
  check("@apollo/client/link/core", linkCore);
  check("@apollo/client/link/error", linkError);
  check("@apollo/client/link/http", linkHTTP);
  check("@apollo/client/link/persisted-queries", linkPersistedQueries);
  check("@apollo/client/link/retry", linkRetry);
  check("@apollo/client/link/schema", linkSchema);
  check("@apollo/client/link/utils", linkUtils);
  check("@apollo/client/link/ws", linkWS);
  check("@apollo/client/react", react);
  check("@apollo/client/react/components", reactComponents);
  check("@apollo/client/react/context", reactContext);
  check("@apollo/client/react/data", reactData);
  check("@apollo/client/react/hoc", reactHOC);
  check("@apollo/client/react/hooks", reactHooks);
  check("@apollo/client/react/parser", reactParser);
  check("@apollo/client/react/ssr", reactSSR);
  check("@apollo/client/testing", testing);
  check("@apollo/client/utilities", utilities);

  it("completeness", () => {
    const { join } = require("path").posix;
    entryPoints.forEach((info: Record<string, any>) => {
      const id = join("@apollo/client", ...info.dirs);
      expect(testedIds).toContain(id);
    });
  });
});
