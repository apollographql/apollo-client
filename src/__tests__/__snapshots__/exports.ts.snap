// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exports of public entry points @apollo/client 1`] = `
Array [
  "ApolloCache",
  "ApolloClient",
  "ApolloConsumer",
  "ApolloError",
  "ApolloLink",
  "ApolloProvider",
  "Cache",
  "DocumentType",
  "HttpLink",
  "InMemoryCache",
  "MissingFieldError",
  "NetworkStatus",
  "Observable",
  "ObservableQuery",
  "checkFetcher",
  "concat",
  "createHttpLink",
  "createSignalIfSupported",
  "defaultDataIdFromObject",
  "disableExperimentalFragmentVariables",
  "disableFragmentWarnings",
  "empty",
  "enableExperimentalFragmentVariables",
  "execute",
  "fallbackHttpConfig",
  "from",
  "fromError",
  "fromPromise",
  "getApolloContext",
  "gql",
  "isApolloError",
  "isReference",
  "makeReference",
  "makeVar",
  "mergeOptions",
  "operationName",
  "parseAndCheckHttpResponse",
  "parser",
  "resetApolloContext",
  "resetCaches",
  "rewriteURIForGET",
  "selectHttpOptionsAndBody",
  "selectURI",
  "serializeFetchParameter",
  "split",
  "throwServerError",
  "toPromise",
  "useApolloClient",
  "useLazyQuery",
  "useMutation",
  "useQuery",
  "useReactiveVar",
  "useSubscription",
]
`;

exports[`exports of public entry points @apollo/client/cache 1`] = `
Array [
  "ApolloCache",
  "Cache",
  "InMemoryCache",
  "MissingFieldError",
  "cacheSlot",
  "defaultDataIdFromObject",
  "isReference",
  "makeReference",
  "makeVar",
]
`;

exports[`exports of public entry points @apollo/client/core 1`] = `
Array [
  "ApolloCache",
  "ApolloClient",
  "ApolloError",
  "ApolloLink",
  "Cache",
  "HttpLink",
  "InMemoryCache",
  "MissingFieldError",
  "NetworkStatus",
  "Observable",
  "ObservableQuery",
  "checkFetcher",
  "concat",
  "createHttpLink",
  "createSignalIfSupported",
  "defaultDataIdFromObject",
  "disableExperimentalFragmentVariables",
  "disableFragmentWarnings",
  "empty",
  "enableExperimentalFragmentVariables",
  "execute",
  "fallbackHttpConfig",
  "from",
  "fromError",
  "fromPromise",
  "gql",
  "isApolloError",
  "isReference",
  "makeReference",
  "makeVar",
  "mergeOptions",
  "parseAndCheckHttpResponse",
  "resetCaches",
  "rewriteURIForGET",
  "selectHttpOptionsAndBody",
  "selectURI",
  "serializeFetchParameter",
  "split",
  "throwServerError",
  "toPromise",
]
`;

exports[`exports of public entry points @apollo/client/errors 1`] = `
Array [
  "ApolloError",
  "isApolloError",
]
`;

exports[`exports of public entry points @apollo/client/link/batch 1`] = `
Array [
  "BatchLink",
  "OperationBatcher",
]
`;

exports[`exports of public entry points @apollo/client/link/batch-http 1`] = `
Array [
  "BatchHttpLink",
]
`;

exports[`exports of public entry points @apollo/client/link/context 1`] = `
Array [
  "setContext",
]
`;

exports[`exports of public entry points @apollo/client/link/core 1`] = `
Array [
  "ApolloLink",
  "concat",
  "empty",
  "execute",
  "from",
  "split",
]
`;

exports[`exports of public entry points @apollo/client/link/error 1`] = `
Array [
  "ErrorLink",
  "onError",
]
`;

exports[`exports of public entry points @apollo/client/link/http 1`] = `
Array [
  "HttpLink",
  "checkFetcher",
  "createHttpLink",
  "createSignalIfSupported",
  "fallbackHttpConfig",
  "parseAndCheckHttpResponse",
  "rewriteURIForGET",
  "selectHttpOptionsAndBody",
  "selectURI",
  "serializeFetchParameter",
]
`;

exports[`exports of public entry points @apollo/client/link/persisted-queries 1`] = `
Array [
  "VERSION",
  "createPersistedQueryLink",
]
`;

exports[`exports of public entry points @apollo/client/link/retry 1`] = `
Array [
  "RetryLink",
]
`;

exports[`exports of public entry points @apollo/client/link/schema 1`] = `
Array [
  "SchemaLink",
]
`;

exports[`exports of public entry points @apollo/client/link/utils 1`] = `
Array [
  "createOperation",
  "fromError",
  "fromPromise",
  "throwServerError",
  "toPromise",
  "transformOperation",
  "validateOperation",
]
`;

exports[`exports of public entry points @apollo/client/link/ws 1`] = `
Array [
  "WebSocketLink",
]
`;

exports[`exports of public entry points @apollo/client/react 1`] = `
Array [
  "ApolloConsumer",
  "ApolloProvider",
  "DocumentType",
  "getApolloContext",
  "operationName",
  "parser",
  "resetApolloContext",
  "useApolloClient",
  "useLazyQuery",
  "useMutation",
  "useQuery",
  "useReactiveVar",
  "useSubscription",
]
`;

exports[`exports of public entry points @apollo/client/react/components 1`] = `
Array [
  "Mutation",
  "Query",
  "Subscription",
]
`;

exports[`exports of public entry points @apollo/client/react/context 1`] = `
Array [
  "ApolloConsumer",
  "ApolloProvider",
  "getApolloContext",
  "resetApolloContext",
]
`;

exports[`exports of public entry points @apollo/client/react/data 1`] = `
Array [
  "MutationData",
  "OperationData",
  "QueryData",
  "SubscriptionData",
]
`;

exports[`exports of public entry points @apollo/client/react/hoc 1`] = `
Array [
  "graphql",
  "withApollo",
  "withMutation",
  "withQuery",
  "withSubscription",
]
`;

exports[`exports of public entry points @apollo/client/react/hooks 1`] = `
Array [
  "useApolloClient",
  "useLazyQuery",
  "useMutation",
  "useQuery",
  "useReactiveVar",
  "useSubscription",
]
`;

exports[`exports of public entry points @apollo/client/react/parser 1`] = `
Array [
  "DocumentType",
  "operationName",
  "parser",
]
`;

exports[`exports of public entry points @apollo/client/react/ssr 1`] = `
Array [
  "RenderPromises",
  "getDataFromTree",
  "getMarkupFromTree",
  "renderToStringWithData",
]
`;

exports[`exports of public entry points @apollo/client/testing 1`] = `
Array [
  "MockLink",
  "MockSubscriptionLink",
  "MockedProvider",
  "createMockClient",
  "itAsync",
  "mockObservableLink",
  "mockSingleLink",
  "stripSymbols",
  "subscribeAndCount",
]
`;

exports[`exports of public entry points @apollo/client/utilities 1`] = `
Array [
  "Concast",
  "DeepMerger",
  "Observable",
  "addTypenameToDocument",
  "argumentsObjectFromField",
  "asyncMap",
  "buildQueryFromSelectionSet",
  "canUseWeakMap",
  "checkDocument",
  "cloneDeep",
  "compact",
  "concatPagination",
  "createFragmentMap",
  "getDefaultValues",
  "getDirectiveNames",
  "getFragmentDefinition",
  "getFragmentDefinitions",
  "getFragmentFromSelection",
  "getFragmentQueryDocument",
  "getInclusionDirectives",
  "getMainDefinition",
  "getOperationDefinition",
  "getOperationName",
  "getQueryDefinition",
  "getStoreKeyName",
  "getTypenameFromResult",
  "graphQLResultHasError",
  "hasClientExports",
  "hasDirectives",
  "isField",
  "isInlineFragment",
  "isNonEmptyArray",
  "isReference",
  "iterateObserversSafely",
  "makeReference",
  "maybeDeepFreeze",
  "mergeDeep",
  "mergeDeepArray",
  "offsetLimitPagination",
  "relayStylePagination",
  "removeArgumentsFromDocument",
  "removeClientSetsFromDocument",
  "removeConnectionDirectiveFromDocument",
  "removeDirectivesFromDocument",
  "removeFragmentSpreadFromDocument",
  "resultKeyNameFromField",
  "shouldInclude",
  "storeKeyNameFromField",
  "valueToObjectRepresentation",
]
`;
